src/custom/components/Tokens/TokensTableRow.tsx

```diff
@@ -24,13 +35,106 @@ type DataRowParams = {
   balance?: CurrencyAmount<Token> | undefined
   handleBuy: (token: Token) => void
   handleSell: (token: Token) => void
+  closeModal: () => void
+  openTransactionConfirmationModal: (message: string, operationType: OperationType) => void
+  toggleWalletModal: () => void
 }
 
-const DataRow = ({ tokenData, index, balance, handleBuy, handleSell }: DataRowParams) => {
+const DataRow = ({
+  tokenData,
+  index,
+  balance,
+  handleBuy,
+  handleSell,
+  closeModal,
+  openTransactionConfirmationModal,
+  toggleWalletModal,
+}: DataRowParams) => {
+  const { account, chainId } = useActiveWeb3React()
+
   const theme = useTheme()
   const hasBalance = balance?.greaterThan(0)
   const formattedBalance = formatSmart(balance) || 0
 
+  // allowance
+  const spender = chainId ? GP_VAULT_RELAYER[chainId] : undefined
+  const currentAllowance = useTokenAllowance(tokenData, account ?? undefined, spender)
+
+  // approve
+  const [approving, setApproving] = useState(false)
+
+  const { handleSetError, handleCloseError } = useErrorModal()
+
+  const { approvalState, approve } = useApproveCallbackFromBalance({
+    openTransactionConfirmationModal,
+    closeModals: closeModal,
+    token: tokenData,
+    balance,
+  })
+
+  const prevApprovalState = usePrevious(approvalState)
+
+  const handleApprove = useCallback(async () => {
+    handleCloseError()
+
+    if (!account) {
+      toggleWalletModal()
+      return
+    }
+
+    try {
+      setApproving(true)
+      const summary = `Approve ${tokenData?.symbol || 'token'}`
+      await approve({ modalMessage: summary, transactionSummary: summary })
+    } catch (error) {
+      console.error(`[TokensTableRow]: Issue approving.`, error)
+      handleSetError(error?.message)
+    } finally {
+      setApproving(false)
+    }
+  }, [account, approve, handleCloseError, handleSetError, toggleWalletModal, tokenData?.symbol])
+
+  const isApproved = approvalState === ApprovalState.APPROVED
+  const isPendingOnchainApprove = approvalState === ApprovalState.PENDING
+  const isPendingApprove = !isApproved && (approving || isPendingOnchainApprove)
+
+  const noBalance = !balance || balance?.equalTo(0)
+
+  const displayApproveContent = useMemo(() => {
+    if (noBalance) {
+      return <ApproveLabel>No balance</ApproveLabel>
+    } else if (isPendingApprove) {
+      return <CardsSpinner />
+    } else if (isApproved) {
+      return <ApproveLabel color={theme.green1}>Approved ✓</ApproveLabel>
+    } else if (!isApproved && currentAllowance && !currentAllowance?.equalTo(0)) {
+      return (
+        <CustomLimit>
+          <TableButton text onClick={handleApprove} color={theme.primary1}>
+            Approve all
+          </TableButton>
+          <ApproveLabel color={theme.green1} title={`Approved: ${currentAllowance.toExact()}`}>
+            Approved: <strong>{formatSmart(currentAllowance, 4)}</strong>
```

Comment:
magic number 4 here. we have PRECISION consts somewhere

Reply:
this will show an amount for tokens with max approval set right? seems overkill. maybe there's a better way to show this. If the amount approved = the max just show `Approved ✅ `

Reply:
Basically like this
![Screenshot from 2022-06-14 16-06-55](https://user-images.githubusercontent.com/34926005/173597437-260d5f72-8be4-4cd4-b032-56f15c5df5b5.png)